engine/src/test/java/io/camunda/zeebe/engine/processing/job/FailJobTest.java

```diff
@@ -272,4 +273,34 @@ public void shouldRejectFailIfErrorThrown() {
     Assertions.assertThat(jobRecord).hasRejectionType(RejectionType.INVALID_STATE);
     assertThat(jobRecord.getRejectionReason()).contains("is in state 'ERROR_THROWN'");
   }
+
+  @Test
+  public void shouldFailWithVariables() {
+    // given
+    ENGINE.createJob(jobType, PROCESS_ID);
+    final Record<JobBatchRecordValue> batchRecord = ENGINE.jobs().withType(jobType).activate();
+    final JobRecordValue job = batchRecord.getValue().getJobs().get(0);
+    final long jobKey = batchRecord.getValue().getJobKeys().get(0);
+    final int retries = 23;
+
+    // when
+    final Record<JobRecordValue> failRecord =
+        ENGINE
+            .job()
+            .withKey(jobKey)
+            .ofInstance(job.getProcessInstanceKey())
+            .withRetries(retries)
+            .withVariables("{'foo':'bar'}")
+            .fail();
+
+    // then
+    Assertions.assertThat(failRecord).hasRecordType(RecordType.EVENT).hasIntent(FAILED);
+    Assertions.assertThat(failRecord.getValue())
+        .hasWorker(job.getWorker())
+        .hasType(job.getType())
+        .hasRetries(retries)
+        .hasDeadline(job.getDeadline());
+
+    assertThat(failRecord.getValue().getVariables()).containsExactly(entry("foo", "bar"));
```

Comment:
❌ This test should also verify that the Variable `foo` is `CREATED` with value `bar` at the variable scope (`failRecord.getValue().getElementInstanceKey()`).

💡 Tip: You can use the `RecordingExporting.variableRecords()` to find the exported variable records.